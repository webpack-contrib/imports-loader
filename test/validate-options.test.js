import { getCompiler, compile } from "./helpers";

describe("validate options", () => {
  const tests = {
    type: {
      success: ["module", "commonjs"],
      failure: ["string", "", {}, []],
    },
    imports: {
      success: [
        "lib_1",
        "globalObject1.foo",
        ["globalObject1"],
        ["globalObject1.foo"],
        {
          moduleName: "jQuery",
          name: "$",
        },
        {
          syntax: "default",
          moduleName: "jQuery",
          name: "lib",
        },
        {
          syntax: "named",
          moduleName: "jQuery",
          name: "lib",
        },
        {
          syntax: "named",
          moduleName: "jQuery",
          name: "lib",
          alias: "lib_alias",
        },
        {
          syntax: "namespace",
          moduleName: "jQuery",
          name: "$",
        },
        {
          syntax: "side-effects",
          moduleName: "jQuery",
        },
        {
          syntax: "single",
          moduleName: "jQuery",
          name: "lib",
        },
        {
          syntax: "multiple",
          moduleName: "jQuery",
          name: "lib",
          alias: "lib_alias",
        },
        {
          syntax: "multiple",
          moduleName: "jQuery",
          name: "lib",
        },
        {
          syntax: "pure",
          moduleName: "jQuery",
        },
      ],
      failure: [
        false,
        true,
        /test/,
        "",
        [],
        [""],
        {},
        {
          type: "string",
          moduleName: "jQuery",
          list: false,
        },
        {
          syntax: "default",
          moduleName: "jQuery",
          name: "lib",
          alias: "lib_alias",
        },
      ],
    },
    wrapper: {
      success: [
        true,
        false,
        "window",
        { thisArg: "window" },
        { thisArg: "window", args: ["foo", "bar"] },
        { thisArg: "window", args: { foo: "bar" } },
      ],
      failure: [
        [],
        [""],
        /test/,
        {},
        { unknown: true },
        { thisArg: 1 },
        { thisArg: "window", args: true },
        { thisArg: "window", args: [1, "bar"] },
      ],
    },
    additionalCode: {
      success: ["var x = 2;"],
      failure: [false, true, /test/, [], [""], {}],
    },
    unknown: {
      success: [],
      failure: [1, true, false, "test", /test/, [], {}, { foo: "bar" }],
    },
  };

  function stringifyValue(value) {
    if (
      Array.isArray(value) ||
      (value && typeof value === "object" && value.constructor === Object)
    ) {
      return JSON.stringify(value);
    }

    return value;
  }

  async function createTestCase(key, value, type) {
    it(`should ${
      type === "success" ? "successfully validate" : "throw an error on"
    } the "${key}" option with "${stringifyValue(value)}" value`, async () => {
      let compiler;

      if (key === "type") {
        compiler = getCompiler("some-library.js", {
          [key]: value,
          wrapper: "window",
        });
      } else {
        compiler = getCompiler("some-library.js", {
          [key]: value,
        });
      }

      let stats;

      try {
        stats = await compile(compiler);
      } finally {
        if (type === "success") {
          const validationErrors = [];

          stats.compilation.errors.forEach((error) => {
            if (error.message.indexOf("ValidationError") !== -1) {
              validationErrors.push(error);
            }
          });

          expect(validationErrors.length).toBe(0);
        } else if (type === "failure") {
          const {
            compilation: { errors },
          } = stats;

          expect(errors).toHaveLength(1);
          expect(() => {
            throw new Error(errors[0].error.message);
          }).toThrowErrorMatchingSnapshot();
        }
      }
    });
  }

  for (const [key, values] of Object.entries(tests)) {
    for (const type of Object.keys(values)) {
      for (const value of values[type]) {
        createTestCase(key, value, type);
      }
    }
  }
});
