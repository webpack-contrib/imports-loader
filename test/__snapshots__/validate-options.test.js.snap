// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`validate options should throw an error on the "additionalCode" option with "/test/" value 1`] = `
"Invalid options object. Imports Loader has been initialized using an options object that does not match the API schema.
 - options.additionalCode should be a non-empty string.
   -> Adds custom code as a preamble before the module's code.
   -> Read more at https://webpack.js.org/loaders/imports-loader/#additionalcode"
`;

exports[`validate options should throw an error on the "additionalCode" option with "[""]" value 1`] = `
"Invalid options object. Imports Loader has been initialized using an options object that does not match the API schema.
 - options.additionalCode should be a non-empty string.
   -> Adds custom code as a preamble before the module's code.
   -> Read more at https://webpack.js.org/loaders/imports-loader/#additionalcode"
`;

exports[`validate options should throw an error on the "additionalCode" option with "[]" value 1`] = `
"Invalid options object. Imports Loader has been initialized using an options object that does not match the API schema.
 - options.additionalCode should be a non-empty string.
   -> Adds custom code as a preamble before the module's code.
   -> Read more at https://webpack.js.org/loaders/imports-loader/#additionalcode"
`;

exports[`validate options should throw an error on the "additionalCode" option with "{}" value 1`] = `
"Invalid options object. Imports Loader has been initialized using an options object that does not match the API schema.
 - options.additionalCode should be a non-empty string.
   -> Adds custom code as a preamble before the module's code.
   -> Read more at https://webpack.js.org/loaders/imports-loader/#additionalcode"
`;

exports[`validate options should throw an error on the "additionalCode" option with "false" value 1`] = `
"Invalid options object. Imports Loader has been initialized using an options object that does not match the API schema.
 - options.additionalCode should be a non-empty string.
   -> Adds custom code as a preamble before the module's code.
   -> Read more at https://webpack.js.org/loaders/imports-loader/#additionalcode"
`;

exports[`validate options should throw an error on the "additionalCode" option with "true" value 1`] = `
"Invalid options object. Imports Loader has been initialized using an options object that does not match the API schema.
 - options.additionalCode should be a non-empty string.
   -> Adds custom code as a preamble before the module's code.
   -> Read more at https://webpack.js.org/loaders/imports-loader/#additionalcode"
`;

exports[`validate options should throw an error on the "imports" option with "" value 1`] = `
"Invalid options object. Imports Loader has been initialized using an options object that does not match the API schema.
 - options.imports should be one of these:
   non-empty string | non-empty string | object { moduleName, syntax?, name?, alias? } | [non-empty string | object { moduleName, syntax?, name?, alias? }, ...] (should not have fewer than 1 item)
   Details:
    * options.imports should be a non-empty string.
      -> Allows to use a string to describe an import.
      -> Read more at https://webpack.js.org/loaders/imports-loader/#string
    * options.imports should be a non-empty string.
      -> Allows to use a string to describe an import.
      -> Read more at https://webpack.js.org/loaders/imports-loader/#string"
`;

exports[`validate options should throw an error on the "imports" option with "/test/" value 1`] = `
"Invalid options object. Imports Loader has been initialized using an options object that does not match the API schema.
 - options.imports misses the property 'moduleName'. Should be:
   non-empty string"
`;

exports[`validate options should throw an error on the "imports" option with "[""]" value 1`] = `
"Invalid options object. Imports Loader has been initialized using an options object that does not match the API schema.
 - options.imports[0] should be a non-empty string.
   -> Allows to use a string to describe an import.
   -> Read more at https://webpack.js.org/loaders/imports-loader/#string"
`;

exports[`validate options should throw an error on the "imports" option with "[]" value 1`] = `
"Invalid options object. Imports Loader has been initialized using an options object that does not match the API schema.
 - options.imports should be a non-empty array."
`;

exports[`validate options should throw an error on the "imports" option with "{"syntax":"default","moduleName":"jQuery","name":"lib","alias":"lib_alias"}" value 1`] = `
"The "default" syntax does not support "lib_alias" alias in "
{
 "syntax": "default",
 "moduleName": "jQuery",
 "name": "lib",
 "alias": "lib_alias"
}
" value"
`;

exports[`validate options should throw an error on the "imports" option with "{"type":"string","moduleName":"jQuery","list":false}" value 1`] = `
"Invalid options object. Imports Loader has been initialized using an options object that does not match the API schema.
 - options.imports should be one of these:
   non-empty string | non-empty string | object { moduleName, syntax?, name?, alias? } | [non-empty string | object { moduleName, syntax?, name?, alias? }, ...] (should not have fewer than 1 item)
   Details:
    * options.imports has an unknown property 'type'. These properties are valid:
      object { moduleName, syntax?, name?, alias? }
      -> Allows to use an object to describe an import.
      -> Read more at https://webpack.js.org/loaders/imports-loader/#object
    * options.imports has an unknown property 'list'. These properties are valid:
      object { moduleName, syntax?, name?, alias? }
      -> Allows to use an object to describe an import.
      -> Read more at https://webpack.js.org/loaders/imports-loader/#object"
`;

exports[`validate options should throw an error on the "imports" option with "{}" value 1`] = `
"Invalid options object. Imports Loader has been initialized using an options object that does not match the API schema.
 - options.imports misses the property 'moduleName'. Should be:
   non-empty string"
`;

exports[`validate options should throw an error on the "imports" option with "false" value 1`] = `
"Invalid options object. Imports Loader has been initialized using an options object that does not match the API schema.
 - options.imports should be one of these:
   non-empty string | non-empty string | object { moduleName, syntax?, name?, alias? } | [non-empty string | object { moduleName, syntax?, name?, alias? }, ...] (should not have fewer than 1 item)
   Details:
    * options.imports should be one of these:
      non-empty string | object { moduleName, syntax?, name?, alias? }
      Details:
       * options.imports should be a non-empty string.
         -> Allows to use a string to describe an import.
         -> Read more at https://webpack.js.org/loaders/imports-loader/#string
       * options.imports should be a non-empty string.
         -> Allows to use a string to describe an import.
         -> Read more at https://webpack.js.org/loaders/imports-loader/#string
       * options.imports should be an object:
         object { moduleName, syntax?, name?, alias? }
         -> Allows to use an object to describe an import.
         -> Read more at https://webpack.js.org/loaders/imports-loader/#object
    * options.imports should be an array:
      [non-empty string | object { moduleName, syntax?, name?, alias? }, ...] (should not have fewer than 1 item)"
`;

exports[`validate options should throw an error on the "imports" option with "true" value 1`] = `
"Invalid options object. Imports Loader has been initialized using an options object that does not match the API schema.
 - options.imports should be one of these:
   non-empty string | non-empty string | object { moduleName, syntax?, name?, alias? } | [non-empty string | object { moduleName, syntax?, name?, alias? }, ...] (should not have fewer than 1 item)
   Details:
    * options.imports should be one of these:
      non-empty string | object { moduleName, syntax?, name?, alias? }
      Details:
       * options.imports should be a non-empty string.
         -> Allows to use a string to describe an import.
         -> Read more at https://webpack.js.org/loaders/imports-loader/#string
       * options.imports should be a non-empty string.
         -> Allows to use a string to describe an import.
         -> Read more at https://webpack.js.org/loaders/imports-loader/#string
       * options.imports should be an object:
         object { moduleName, syntax?, name?, alias? }
         -> Allows to use an object to describe an import.
         -> Read more at https://webpack.js.org/loaders/imports-loader/#object
    * options.imports should be an array:
      [non-empty string | object { moduleName, syntax?, name?, alias? }, ...] (should not have fewer than 1 item)"
`;

exports[`validate options should throw an error on the "type" option with "" value 1`] = `
"Invalid options object. Imports Loader has been initialized using an options object that does not match the API schema.
 - options.type should be one of these:
   "module" | "commonjs"
   -> Format of generated exports.
   -> Read more at https://github.com/webpack-contrib/imports-loader#type"
`;

exports[`validate options should throw an error on the "type" option with "[]" value 1`] = `
"Invalid options object. Imports Loader has been initialized using an options object that does not match the API schema.
 - options.type should be one of these:
   "module" | "commonjs"
   -> Format of generated exports.
   -> Read more at https://github.com/webpack-contrib/imports-loader#type"
`;

exports[`validate options should throw an error on the "type" option with "{}" value 1`] = `
"Invalid options object. Imports Loader has been initialized using an options object that does not match the API schema.
 - options.type should be one of these:
   "module" | "commonjs"
   -> Format of generated exports.
   -> Read more at https://github.com/webpack-contrib/imports-loader#type"
`;

exports[`validate options should throw an error on the "type" option with "string" value 1`] = `
"Invalid options object. Imports Loader has been initialized using an options object that does not match the API schema.
 - options.type should be one of these:
   "module" | "commonjs"
   -> Format of generated exports.
   -> Read more at https://github.com/webpack-contrib/imports-loader#type"
`;

exports[`validate options should throw an error on the "unknown" option with "/test/" value 1`] = `
"Invalid options object. Imports Loader has been initialized using an options object that does not match the API schema.
 - options should be one of these:
   object { imports, … } | object { wrapper, … } | object { additionalCode, … }
   -> Options for imports-loader
   Details:
    * options has an unknown property 'unknown'. These properties are valid:
      object { imports, … } | object { wrapper, … } | object { additionalCode, … }
      -> Options for imports-loader
    * options misses the property 'imports' | should be any non-object.
    * options misses the property 'wrapper' | should be any non-object.
    * options misses the property 'additionalCode' | should be any non-object."
`;

exports[`validate options should throw an error on the "unknown" option with "[]" value 1`] = `
"Invalid options object. Imports Loader has been initialized using an options object that does not match the API schema.
 - options should be one of these:
   object { imports, … } | object { wrapper, … } | object { additionalCode, … }
   -> Options for imports-loader
   Details:
    * options has an unknown property 'unknown'. These properties are valid:
      object { imports, … } | object { wrapper, … } | object { additionalCode, … }
      -> Options for imports-loader
    * options misses the property 'imports' | should be any non-object.
    * options misses the property 'wrapper' | should be any non-object.
    * options misses the property 'additionalCode' | should be any non-object."
`;

exports[`validate options should throw an error on the "unknown" option with "{"foo":"bar"}" value 1`] = `
"Invalid options object. Imports Loader has been initialized using an options object that does not match the API schema.
 - options should be one of these:
   object { imports, … } | object { wrapper, … } | object { additionalCode, … }
   -> Options for imports-loader
   Details:
    * options has an unknown property 'unknown'. These properties are valid:
      object { imports, … } | object { wrapper, … } | object { additionalCode, … }
      -> Options for imports-loader
    * options misses the property 'imports' | should be any non-object.
    * options misses the property 'wrapper' | should be any non-object.
    * options misses the property 'additionalCode' | should be any non-object."
`;

exports[`validate options should throw an error on the "unknown" option with "{}" value 1`] = `
"Invalid options object. Imports Loader has been initialized using an options object that does not match the API schema.
 - options should be one of these:
   object { imports, … } | object { wrapper, … } | object { additionalCode, … }
   -> Options for imports-loader
   Details:
    * options has an unknown property 'unknown'. These properties are valid:
      object { imports, … } | object { wrapper, … } | object { additionalCode, … }
      -> Options for imports-loader
    * options misses the property 'imports' | should be any non-object.
    * options misses the property 'wrapper' | should be any non-object.
    * options misses the property 'additionalCode' | should be any non-object."
`;

exports[`validate options should throw an error on the "unknown" option with "1" value 1`] = `
"Invalid options object. Imports Loader has been initialized using an options object that does not match the API schema.
 - options should be one of these:
   object { imports, … } | object { wrapper, … } | object { additionalCode, … }
   -> Options for imports-loader
   Details:
    * options has an unknown property 'unknown'. These properties are valid:
      object { imports, … } | object { wrapper, … } | object { additionalCode, … }
      -> Options for imports-loader
    * options misses the property 'imports' | should be any non-object.
    * options misses the property 'wrapper' | should be any non-object.
    * options misses the property 'additionalCode' | should be any non-object."
`;

exports[`validate options should throw an error on the "unknown" option with "false" value 1`] = `
"Invalid options object. Imports Loader has been initialized using an options object that does not match the API schema.
 - options should be one of these:
   object { imports, … } | object { wrapper, … } | object { additionalCode, … }
   -> Options for imports-loader
   Details:
    * options has an unknown property 'unknown'. These properties are valid:
      object { imports, … } | object { wrapper, … } | object { additionalCode, … }
      -> Options for imports-loader
    * options misses the property 'imports' | should be any non-object.
    * options misses the property 'wrapper' | should be any non-object.
    * options misses the property 'additionalCode' | should be any non-object."
`;

exports[`validate options should throw an error on the "unknown" option with "test" value 1`] = `
"Invalid options object. Imports Loader has been initialized using an options object that does not match the API schema.
 - options should be one of these:
   object { imports, … } | object { wrapper, … } | object { additionalCode, … }
   -> Options for imports-loader
   Details:
    * options has an unknown property 'unknown'. These properties are valid:
      object { imports, … } | object { wrapper, … } | object { additionalCode, … }
      -> Options for imports-loader
    * options misses the property 'imports' | should be any non-object.
    * options misses the property 'wrapper' | should be any non-object.
    * options misses the property 'additionalCode' | should be any non-object."
`;

exports[`validate options should throw an error on the "unknown" option with "true" value 1`] = `
"Invalid options object. Imports Loader has been initialized using an options object that does not match the API schema.
 - options should be one of these:
   object { imports, … } | object { wrapper, … } | object { additionalCode, … }
   -> Options for imports-loader
   Details:
    * options has an unknown property 'unknown'. These properties are valid:
      object { imports, … } | object { wrapper, … } | object { additionalCode, … }
      -> Options for imports-loader
    * options misses the property 'imports' | should be any non-object.
    * options misses the property 'wrapper' | should be any non-object.
    * options misses the property 'additionalCode' | should be any non-object."
`;

exports[`validate options should throw an error on the "wrapper" option with "/test/" value 1`] = `
"Invalid options object. Imports Loader has been initialized using an options object that does not match the API schema.
 - options.wrapper misses the property 'thisArg'. Should be:
   non-empty string"
`;

exports[`validate options should throw an error on the "wrapper" option with "[""]" value 1`] = `
"Invalid options object. Imports Loader has been initialized using an options object that does not match the API schema.
 - options.wrapper should be one of these:
   boolean | non-empty string | object { thisArg, args? }
   -> Closes the module code in a function with a given 'thisArg' and 'args'
   -> Read more at https://webpack.js.org/loaders/imports-loader/#wrapper
   Details:
    * options.wrapper should be a boolean.
    * options.wrapper should be a non-empty string.
    * options.wrapper should be an object:
      object { thisArg, args? }"
`;

exports[`validate options should throw an error on the "wrapper" option with "[]" value 1`] = `
"Invalid options object. Imports Loader has been initialized using an options object that does not match the API schema.
 - options.wrapper should be one of these:
   boolean | non-empty string | object { thisArg, args? }
   -> Closes the module code in a function with a given 'thisArg' and 'args'
   -> Read more at https://webpack.js.org/loaders/imports-loader/#wrapper
   Details:
    * options.wrapper should be a boolean.
    * options.wrapper should be a non-empty string.
    * options.wrapper should be an object:
      object { thisArg, args? }"
`;

exports[`validate options should throw an error on the "wrapper" option with "{"thisArg":"window","args":[1,"bar"]}" value 1`] = `
"Invalid options object. Imports Loader has been initialized using an options object that does not match the API schema.
 - options.wrapper.args[0] should be a non-empty string."
`;

exports[`validate options should throw an error on the "wrapper" option with "{"thisArg":"window","args":true}" value 1`] = `
"Invalid options object. Imports Loader has been initialized using an options object that does not match the API schema.
 - options.wrapper should be one of these:
   boolean | non-empty string | object { thisArg, args? }
   -> Closes the module code in a function with a given 'thisArg' and 'args'
   -> Read more at https://webpack.js.org/loaders/imports-loader/#wrapper
   Details:
    * options.wrapper.args should be one of these:
      [non-empty string, ...] (should not have fewer than 1 item) | object { … }
      Details:
       * options.wrapper.args should be an array:
         [non-empty string, ...] (should not have fewer than 1 item)
       * options.wrapper.args should be an object:
         object { … }"
`;

exports[`validate options should throw an error on the "wrapper" option with "{"thisArg":1}" value 1`] = `
"Invalid options object. Imports Loader has been initialized using an options object that does not match the API schema.
 - options.wrapper.thisArg should be a non-empty string."
`;

exports[`validate options should throw an error on the "wrapper" option with "{"unknown":true}" value 1`] = `
"Invalid options object. Imports Loader has been initialized using an options object that does not match the API schema.
 - options.wrapper should be one of these:
   boolean | non-empty string | object { thisArg, args? }
   -> Closes the module code in a function with a given 'thisArg' and 'args'
   -> Read more at https://webpack.js.org/loaders/imports-loader/#wrapper
   Details:
    * options.wrapper has an unknown property 'unknown'. These properties are valid:
      object { thisArg, args? }
    * options.wrapper misses the property 'thisArg'. Should be:
      non-empty string"
`;

exports[`validate options should throw an error on the "wrapper" option with "{}" value 1`] = `
"Invalid options object. Imports Loader has been initialized using an options object that does not match the API schema.
 - options.wrapper misses the property 'thisArg'. Should be:
   non-empty string"
`;
